### PR TITLE
[CALCITE-4159] Simplify always-true expressions (such as LIKE '%') to TRUE

### PR DESCRIPTION
**Pull request by julianhyde**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 1]
```diff
@@ -2284,6 +2291,12 @@ static Predicate of(RexNode t) {
       }
       return IsPredicate.of(t);
     }
+
+    /** Returns whether this predicate can while simplifying other OR
+     * operands. */
+    default boolean allowedInOr(RelOptPredicateList predicates) {
+      return true;
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java:2298]: It seems to lack some words between `can` and `while`.

File: core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml [hunk 2]
```diff
@@ -3763,7 +3765,7 @@ LogicalProject(DEPTNO=[$7])
     </TestCase>
     <TestCase name="testReduceConstExpr">
         <Resource name="sql">
-            <![CDATA[select sum(case when 'y' = 'n' then ename else 1 end) from emp]]>
+            <![CDATA[select sum(case when 'y' = 'n' then ename else 0.1 end) from emp]]>
         </Resource>
```
Comment (chunweilei) [inline core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:3769]: Is this change correct?

File: core/src/test/java/org/apache/calcite/rex/RexProgramBuilderBase.java [hunk 3]
```diff
@@ -306,6 +306,14 @@ protected RexNode gt(RexNode n1, RexNode n2) {
...
   }
 
+  protected RexNode like(RexNode ref, RexNode pattern) {
+    return rexBuilder.makeCall(SqlStdOperatorTable.LIKE, ref, pattern);
+  }
+
+  protected RexNode like(RexNode ref, RexNode pattern, RexNode escape) {
+    return rexBuilder.makeCall(SqlStdOperatorTable.LIKE, ref, pattern, escape);
+  }
+
```
Comment (chunweilei) [inline core/src/test/java/org/apache/calcite/rex/RexProgramBuilderBase.java:316]: What does this method mean? Could you add some comments?

File: core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml (hunk 2 above)
Comment (amaliujia) [reply core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:3769]: +1. The test case should be using `1`: https://github.com/apache/calcite/blob/37b8cdbb381369e773229a81ae69ab5c1df34f3e/core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java#L2649

Comment (chunweilei) [reply core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:3769]: Got it, 3q.

Comment (julianhyde) [reply core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:3769]: Yes, the 'sql' originates in `SqlToRelConverterTest.java`. I just updated `SqlToRelConverterTest.xml` to match what was already in the java.

File: core/src/test/java/org/apache/calcite/rex/RexProgramBuilderBase.java (hunk 3 above)
Comment (julianhyde) [reply core/src/test/java/org/apache/calcite/rex/RexProgramBuilderBase.java:316]: I disagree. There are lots of surrounding methods with a similar pattern. They are protected methods in a test case. Javadoc would not add anything.

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java (hunk 1 above)
Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java:2298]: good catch. i'll fix.

File: core/src/main/java/org/apache/calcite/plan/RelOptPredicateList.java [hunk 4]
```diff
@@ -180,4 +203,27 @@ public RelOptPredicateList shift(RexBuilder rexBuilder, int offset) {
...
+          && ((RexCall) p).getOperands().get(0).equals(e)) {
+        return true;
+      }
+    }
+    if (SqlKind.COMPARISON.contains(e.getKind())) {
+      // A comparison with a literal, such as 'ref < 10', is not null if 'ref'
+      // is not null.
+      RexCall call = (RexCall) e;
+      if (call.getOperands().get(1) instanceof RexLiteral) {
+        return isEffectivelyNotNull(call.getOperands().get(0));
```
Comment (liyafan82) [inline core/src/main/java/org/apache/calcite/plan/RelOptPredicateList.java:224]: do we also need to consider cases such as '10 > ref' here?

**Pull request closed without merging**